gr-digital/lib/constellation.cc

```diff
@@ -269,10 +319,10 @@ std::vector<float> constellation::calc_soft_dec(gr_complex sample, float npwr)
     for (int i = 0; i < M; i++) {
         // Calculate the distance between the sample and the current
         // constellation point.
-        float dist = std::abs(sample - d_constellation[i]);
+        float dist = powf(std::abs(sample - d_constellation[i]), 2.0f);
         // Calculate the probability factor from the distance and
-        // the scaled noise power.
-        float d = expf(-dist / npwr);
+        // the scaled noise power. (divisor 2.0f for noise amplitude, 1.0f for power?)
+        float d = expf(-dist / (d_npwr * 1.0f));
```

Comment:
I'm not sure about this. I would expect a `2.0` here.

The probability density function for receiving `sample` conditioned to the fact that the transmitted symbol was `d_constellation[i]` is `exp(-|sample - d_constellation[i]|^2/(2 * sigma^2))`, where `sigma` is the noise standard deviation.

I'd have to check all the math here to be sure, since the way I'm used to computing LLRs for received symbols from a constellation is to compute `a[i] = Re(sample * conj(d_constellation[i])) / sigma^2`, then max* reduce the `a[i]` for those symbols in which the bit has the value 0 (call the result `b`), max* reduce the `a[i]` for those symbols in which the bit has the value 1 (call the result `c`), and subtract `c - b`.

Reply:
Looking more at this code (and comparing with some notes I have for how this can be computed in different ways), I see that what the code is doing is to sum the PDFs of `sample` conditioned to `d_constellation[i]` for those `i`'s for which the bit is 1 (call this `one`), sum the PDFs of `sample` conditioned to `d_constellation[i]` for those `i`'s for which the bit is 0 (call this `zero`), and then computing `log(one) - log(zero)`. This is the way to compute the LLR following the definition.

So now I'm convinced that it should be
```c++
float d = expf(-dist / (d_npwr * 2.0f));
```

Reply:
I would tend to agree, I remember looking at this before. The only thing I don't understand is why this change seems to improve BER. I suppose I could be accidentally calculating my noise power incorrectly. My attached flowgraph illustrates this difference. I calculate my noise power as 10^(-SNR_dB/10) which is the input parameter sigma. Maybe there's a missing power of 2 or square root somewhere.

Reply:
You're right. I was mistaking noise power and noise sigma for complex Gaussian noise. Since `noise_pwr = 2 * sigma^2`, the denominator needs to have `d_npwr`, not `2.0 * d_npwr`.